apps/api/src/core/services/tracing/tracing.service.ts

```diff
@@ -0,0 +1,182 @@
+
+      const span = tracer.startSpan(name);
+
+      try {
+        if (originalMethod.constructor.name === "AsyncFunction" || originalMethod.toString().includes("return __awaiter")) {
+          return originalMethod
+            .apply(this, args)
+            .then((result: any) => {
+              span.setStatus({ code: SpanStatusCode.OK });
+              span.end();
+              return result;
+            })
+            .catch((error: Error) => {
+              span.setStatus({
+                code: SpanStatusCode.ERROR,
+                message: error.message
+              });
+              span.recordException(error);
+              span.end();
+              throw error;
+            });
+        }
+
+        const result = originalMethod.apply(this, args);
+        span.setStatus({ code: SpanStatusCode.OK });
+        span.end();
+        return result;
+      } catch (error) {
+        span.setStatus({
+          code: SpanStatusCode.ERROR,
+          message: error.message
+        });
+        span.recordException(error);
+        span.end();
+        throw error;
+      }
+    };
+
+    return descriptor;
+  };
+}
+
+/**
+ * Utility for manually creating and managing spans
+ * Use this when you need fine-grained control over span lifecycle and attributes.
+ *
+ * @example
+ * ```typescript
+ * async function processItems(items: Item[]) {
+ *   const span = createSpan('batch-processing');
+ *
+ *   try {
+ *     // Add custom attributes to span
+ *     span.setAttribute('items.count', items.length);
+ *     span.setAttribute('process.type', 'bulk');
+ *
+ *     // Process items
+ *     for (const item of items) {
+ *       await processItem(item);
+ *     }
+ *
+ *     // Set success status
+ *     span.setStatus({ code: SpanStatusCode.OK });
+ *   } catch (error) {
+ *     // Record error details
+ *     span.setStatus({
+ *       code: SpanStatusCode.ERROR,
+ *       message: error.message
+ *     });
+ *     span.recordException(error);
+ *     throw error;
+ *   } finally {
+ *     // Always end the span
+ *     span.end();
+ *   }
+ * }
+ * ```
+ *
+ * @param name - Name of the span
+ * @returns OpenTelemetry Span object
+ */
+export function createSpan(name: string) {
+  return trace.getTracer("default").startSpan(name);
+}
+
+/**
+ * Executes the provided function within a new span
+ * Handles span lifecycle automatically, including error cases
+ *
+ * @example
+ * ```typescript
+ * // Simple usage
+ * async function importData() {
+ *   const result = await withSpan('data-import', async () => {
+ *     // Complex operation logic
+ *     const data = await fetchData();
+ *     return processData(data);
+ *   });
+ *
+ *   return result;
+ * }
+ *
+ * // With span attributes
+ * async function processOrder(orderId: string) {
+ *   return withSpan('order-processing', async () => {
+ *     // Get current span via context
+ *     const span = trace.getSpan(context.active());
```

Comment:
Where does the context comes from? Can we pass the current span in the function instead?

```
return withSpan('order-processing', async span => {
   span.setAttribute(...);
}
```

Reply:
Context is exported from otel api package. Any improvements are welcome for sure. This pr is functional though and already covers the use case with benchmark implementation from ur pr.